language/ankush-playground/src/Vote.move

```diff
@@ -0,0 +1,260 @@
+    }
+
+    // This is a helper struct which allows us to store a map of proposal_type -> policy
+    // in the ApprovalPolicies struct
+    struct ApprovalPolicyMap has store, copy {
+        proposal_type: vector<u8>,
+        policy: ApprovalPolicy,
+    }
+
+    // Used to store onchain the current approval policy list. Sample mapping:
+    // [
+    //   "create_validator_owner" -> {allowed_voters: ["alice", "bob", "carol"], num_votes_required: 2},
+    //   "create_designated_dealer" -> {allowed_voters: ["bob", "carol", "dave"], num_votes_required: 1}
+    // ]
+    struct ApprovalPolicies has key {
+        approval_policies: vector<ApprovalPolicyMap>,
+    }
+
+    // An identifier for a ballot. Instead of a global counter,
+    // we use a separate counter under each proposer's address
+    struct BallotID has store, copy, drop {
+        counter: u64,
+        proposer: address,
+    }
+
+    /// A ballot is a struct parametrized on a Proposal.
+    struct Ballot<Proposal: store> has store {
+        // details for the proposal being voted on.
+        proposal: Proposal,
+        // A string type for the proposal - ex: "create_validator_owner",
+        // "freeze_account", "create_vasp", etc.
+        // This lives outside the proposal: Proposal to make it easy for
+        // indexers to index the ballots which have been proposed and
+        // categorize them into "types"
+        proposal_type: vector<u8>,
+        approval_policy: ApprovalPolicy,
+        votes_received: vector<address>,
+        // a globally unique ballot id that is created for every proposal
+        ballot_id: BallotID,
+        // Votes rejected before this time
+        start_timestamp_secs: u64,
+        // Votes rejected after this time
+        expiration_timestamp_secs: u64,
+    }
+
+    struct Ballots<Proposal: store> has key {
+        ballots: vector<Ballot<Proposal>>
+    }
+
+    // A counter which is stored under the proposers address and gets incremented
+    // everytime they create a new ballot
+    struct BallotCounter has key {
+        counter: u64,
+    }
+
+    // Creates an empty list of approval policies stored at the root account
+    public(script) fun initialize_approval_policies(root_account: &signer) {
+        // We store the all the approval policies in the root address, hence this
+        // requirement.
+        // TODO assert_diem_root gives UNEXPECTED_VERIFIER_ERROR
+        // Roles::assert_diem_root(root_account);
+        let address = Signer::address_of(root_account);
+        assert(!exists<ApprovalPolicies>(address), EAPPROVAL_POLICES_EXIST);
+        move_to(root_account, ApprovalPolicies {
+                approval_policies: Vector::empty(),
+        });
+    }
+
+    // store_approval_policy stores the provided approval policy for the given
+    // proposal_type in global storage. If a policy already exists for a given
+    // proposal_type, it updates the policy
+    public fun store_approval_policy(
```

Comment:
I would prefer storing the code for approval policies in a single module so that they can be shared across all of its clients. Although I have not added the functions here, we would need some functions which allow us to make changes to the approval policies (adding/removing voters, updating threshold, etc) of operations and that would be easier if it lives in its own module.

Since Vote and ApprovalPolicy are so closely related to each other, I put them in a single module, but having a separate ApprovalPolicy module also makes sense.